ols/src/query_helpers/docs_summarizer.py

```diff
@@ -123,6 +123,13 @@ def summarize(
             history,
             rag_context,
         )
+
+        # Final check if we don't use more tokens than permitted by provider+model
+        # Handles: OLS-538
+        token_handler.get_available_tokens(
+            final_prompt.format(**llm_input_values), model_config
+        )
```

Comment:
does question_validator have a similar check?  Since the prompt might go there before it ever gets to docs_summarizer

Reply:
on line 64, but it does not catch all issues with total length (no history nor rag)

Reply:
> on line 64, but it does not catch all issues with total length (no history nor rag)

sure, i just wanted to be sure that if the prompt itself (w/o rag+history) is too long, the question validator isn't going to blow up, but rather will also return a clean error.

i don't see that check on line 64 of question_validator currently, but perhaps you're referring to an unmerged pr?

Reply:
it's this line
```
        TokenHandler().get_available_tokens(query, model_config)
```

the function has horrible name as it does two things:
- returns available tokens if > 0
- throws an exception instead

so probably better name would be check_and... or ensure_... or we can split into two functions, but probably it is nitpicking right now

Reply:
yes it's a bit surprising/confusing that "get_available_tokens" is intended to raise an exception.  Even more so confusing because validate_question (the function that's calling get_available_tokens) isn't attempting to catch that exception.  So you're relying on the caller of validate_question to know that validate_question is calling a "get" function that may throw an exception, and the caller of validate_question is expected to handle that exception.

too many layers of indirection.

Reply:
(but i'm glad that the check is being done in validate_question in addition to docs_summarizer)

Reply:
yup, let's create function `ensure_available_tokens(available_tokens, needed_tokens=0)` ATM the `needed_tokens` wont be needed (sic!), but in future might be useful to check against some minimal required value, and not just with 0.